Jet met updates

yiiyama opened the pull request:
Mostly updates on Bambu side. We are temporarily filling both ak8 and ca15 fat jets. Will decide how to proceed in the coming weeks.

To accommodate this multi-collection configuration, defineBranches and setBranchAddresses of the BareFatJets was modified to accept a prefix argument (which will be e.g. ak8 or ca15), with an empty default. The change is transparent to the existing code.

yiiyama commented:
There is another important non-bambu update, this one not transparent: I removed CH, NH, and Ph MET and added the trackMet. In NeroMet this is defined as the negative vector sum of all charged PF candidates. In Bambu MetFiller the definition is the negative vector sum of all CH, e, and mu. The two definitions should be identical.

amarini commented:
* Fancy mets were not used so far from our side.
* Shouldn't we require some mathing to the primary vertex, or better exclude the one matched to other pv vertices (as in beta* and CHS)? 
        * eg, https://github.com/MiT-HEP/NeroProducer/blob/master/Nero/src/NeroJets.cpp#L101
* for the fatjets, I would put choose stable version in the long run. For the moment we are not using them yet (in the MiniAOD path there is a bug in the subjects of these fatjets #44 ). We are mainly interested in fatjets for the Hbb tagger, but we may consider interesting W/top tagging too.

amarini commented on Core/interface/BareP4.hpp:
```diff
@@ -34,8 +34,10 @@ class BareP4 : virtual public BareCollection
+        virtual void defineBranches(TTree* t, std::string prefix);
+        virtual void setBranchAddresses(TTree* t, std::string prefix);
+        void defineBranches(TTree* t) override { defineBranches(t, ""); }
+        void setBranchAddresses(TTree* t) override { setBranchAddresses(t, ""); }
```
These should be virtual, since they are called by BarePhotons, BareLeptons ...

amarini commented on Core/interface/BareFatJets.hpp:
```diff
@@ -12,8 +12,9 @@ class BareFatJets : virtual public BareP4
         void init();
-        void defineBranches(TTree *t);
-        virtual void setBranchAddresses(TTree*);
+        // non-prefix versions are defined in BareP4 and will call the following functions with prefix = ""
```
I'm afraid you have to redefine also the defineBranches(TTree*t) , or overload the next one with prefix="",
otherwise the BareFatJets class will not be able to call defineBranches(t). I think this is related to how the classes inherithance is written from the compiler.
( so , BareFatJets->defineBranches(t) -> ERROR, while static_cast(BareCollection*)<&b>->defineBranches(t)//ok )

amarini commented:
See comment inline. To summurize:
* non prefix member of BareP4 should be virtual, since some derived classes may override only these two.
* BareFatjets and BareJets should define the one without prefix, otherwise these methods will not be callable from the derived classes.

yiiyama commented:
OK, added ```using BareP4::*``` statements to the P4-derived classes.
Also got rid of virtual statements in derived classes and replaced them with override, so that it is more clear where in the inheritance tree a certain function is first declared.

yiiyama commented:
reverting the use of "using" and declared both the prefix and non-prefix versions for all classes.

amarini commented:
thanks, ... It's less elegant ... but works smoothly in python ....
